### ISSUE TITLE
Set boot flag

### ISSUE THREAD

**Issue by aliveafter1000** [calamares/calamares#1046]
https://forum.manjaro.org/t/solved-boot-loop-before-grub/63775/19

https://en.wikipedia.org/wiki/Boot_flag

"Some buggy BIOSes test if the bootflag of at least one partition is set. Otherwise they ignore the device in boot-order. So even if the bootloader does not need the flag, it has to be set to start the bootcode from BIOS.”

Currently, no boot flag is set (automated installation).
Many people have problems. Particularly old laptops.
```

    Partition* rootPartition = nullptr;
    if ( luksPassphrase.isEmpty() )
    {
        rootPartition = KPMHelpers::createNewPartition(
            dev->partitionTable(),
            *dev,
            PartitionRole( PartitionRole::Primary ),
            FileSystem::typeForName( defaultFsType ),
            firstFreeSector,
            lastSectorForRoot
        );
    }
    else
    {
        rootPartition = KPMHelpers::createNewEncryptedPartition(
            dev->partitionTable(),
            *dev,
            PartitionRole( PartitionRole::Primary ),
            FileSystem::typeForName( defaultFsType ),
            firstFreeSector,
            lastSectorForRoot,
            luksPassphrase
       );
    }
    PartitionInfo::setFormat( rootPartition, true );
    PartitionInfo::setMountPoint( rootPartition, "/" );
core->createPartition( dev, rootPartition );
```

**Comment by aliveafter1000** [comment]
Also, 

```
 if ( isEfi )
    {
        qint64 efiSectorCount = bytesToSectors( uefisys_part_sizeB, dev->logicalSize() );
        Q_ASSERT( efiSectorCount > 0 );

        // Since sectors count from 0, and this partition is created starting
        // at firstFreeSector, we need efiSectorCount sectors, numbered
        // firstFreeSector..firstFreeSector+efiSectorCount-1.
        qint64 lastSector = firstFreeSector + efiSectorCount - 1;
        core->createPartitionTable( dev, PartitionTable::gpt );
        Partition* efiPartition = KPMHelpers::createNewPartition(
            dev->partitionTable(),
            *dev,
            PartitionRole( PartitionRole::Primary ),
            FileSystem::Fat32,
            firstFreeSector,
            lastSector,
            PartitionTable::FlagNone
        );
        PartitionInfo::setFormat( efiPartition, true );
        PartitionInfo::setMountPoint( efiPartition, gs->value( "efiSystemPartition" )
                                                        .toString() );
        core->createPartition( dev, efiPartition, PartitionTable::FlagEsp );
        firstFreeSector = lastSector + 1;
}
```
It does not *seem* to need PartitionTable::FlagNone considering that's the default?
https://github.com/calamares/calamares/blob/743c19d8dd58bc7e7ee26b3fca769318d254d172/src/modules/partition/core/KPMHelpers.h
```
Partition* createNewPartition( PartitionNode* parent,
                               const Device& device,
                               const PartitionRole& role,
                               FileSystem::Type fsType,
                               qint64 firstSector,
                               qint64 lastSector,
PartitionTable::Flags flags = PartitionTable::FlagNone );
```
Also confused to why you can set flags in both createNewPartition and CreatePartition?
Just trying to understand.

**Comment by abucodonosor** [comment]
@aliveafter1000 

can you send PR's in please ?

**Comment by aliveafter1000** [comment]
Since most people won't know what's wrong when they fail to boot when they decide to do manual partitioning. You should also add a warning in the manual partition part. 
Warn that it may not boot on some devices if failure to set the boot flag.
I know there are several warnings to help guide users when they manually partition.

**Comment by aliveafter1000** [comment]
@stikonas 
@adriaandegroot

**FlagEsp [[deprecated]] = FlagBoot**

https://github.com/KDE/kpmcore/blob/4080eef0bab6a9334fb6d36f2f337fc43a64797e/src/core/partitiontable.h
```
    /** Partition flags */
    enum Flag : qint32 {
        FlagNone = 0,
        FlagBoot = 1,
        FlagRoot = 2,
        FlagSwap = 4,
        FlagHidden = 8,
        FlagRaid = 16,
        FlagLvm = 32,
        FlagLba = 64,
        FlagHpService = 128,
        FlagPalo = 256,
        FlagPrep = 512,
        FlagMsftReserved = 1024,
        FlagBiosGrub = 2048,
        FlagAppleTvRecovery = 4096,
        FlagDiag = 8192,
        FlagLegacyBoot = 16384,
        FlagMsftData = 32768,
        FlagIrst = 65536,
        FlagEsp [[deprecated]] = FlagBoot
    };
```

If FlagEsp = FlagBoot,
That is a problem for this code.

https://github.com/calamares/calamares/blob/6de55e6951fd0956bcbd44b60714dad686132050/src/modules/partition/core/PartUtils.cpp
```
bool
isEfiBootable( const Partition* candidate )
{
    cDebug() << "Check EFI bootable" << candidate->partitionPath() << candidate->devicePath();
    cDebug() << " .. flags" << candidate->activeFlags();

    auto flags = PartitionInfo::flags( candidate );

    /* If bit 17 is set, old-style Esp flag, it's OK */
    if ( flags.testFlag( PartitionTable::FlagEsp ) )
        return true;

    /* Otherwise, if it's a GPT table, Boot (bit 0) is the same as Esp */
    const PartitionNode* root = candidate;
    while ( root && !root->isRoot() )
    {
        root = root->parent();
        cDebug() << " .. moved towards root" << (void *)root;
    }

    // Strange case: no root found, no partition table node?
    if ( !root )
        return false;

    const PartitionTable* table = dynamic_cast<const PartitionTable*>( root );
    cDebug() << "  .. partition table" << (void *)table << "type" << ( table ? table->type() : PartitionTable::TableType::unknownTableType );
    return table && ( table->type() == PartitionTable::TableType::gpt ) &&
        flags.testFlag( PartitionTable::FlagBoot );
}
```

**Comment by stikonas** [comment]
I just quickly glanced at the code but why is it a problem?
First "return" will return if partition is marked either FlagESP or FlagBoot.

Somehow old kpmcore just had two different flags for the same thing which is confusing,
so I thought I would simplify things.

**Comment by stikonas** [comment]
On Tue, Nov 06, 2018 at 04:25:24AM -0800, aliveafter1000 wrote:
> Because it was written with FlagEsp and FlagBoot being different in mind...
> 
>     /* If bit 17 is set, old-style Esp flag, it's OK */
>     if ( flags.testFlag( PartitionTable::FlagEsp ) )
>         return true;
> 
>     return table && ( table->type() == PartitionTable::TableType::gpt ) &&
>         flags.testFlag( PartitionTable::FlagBoot );

So if FlagBoot=FlatEsp=true, then first return true will work.

If FlagBoot=FlagEsp=false, then first if will be skipped and second will return
false die to flag.testFag (PartitionTable::FlagBoot) returning false.
So I still don't see how it fails? Am I missing something in my reasoning?

**Comment by adriaandegroot** [comment]
@aliveafter1000 also, in general please try to do one thing at a time in an issue.

**Comment by aliveafter1000** [comment]
That's what I'm saying. It has no purpose any more because:
PartitionTable::FlagEsp == PartitionTable::FlagBoot

Right now, it is isBootable()

As for the PR:
EFI partition is bootable.
hence:

```
    if( isEfi )
    {
        core->createPartition( dev, rootPartition );
    }
    else
    {
        // Some buggy BIOSes test if the bootflag of at least one partition is set. Otherwise they ignore the device in boot-order. 
        core->createPartition( dev, rootPartition, PartitionTable::FlagBoot );
    }
```

**Comment by stikonas** [comment]
On Tue, Nov 06, 2018 at 06:33:15AM -0800, aliveafter1000 wrote:
> That's what I'm saying. It has no purpose any more because:
> PartitionTable::FlagEsp == PartitionTable::FlagBoot

Not in older versions of kpmcore. So we need it for compatibility
with older kpmcore versions. In the future it can of course be simplified.
That's why I wanted to simplify it in kpmcore in the first place.

**Comment by philmmanjaro** [comment]
@aliveafter1000, always add both code sections to the upstream code of Calamares. On Manjaro we always use the latest code and libraries, however other distros like Debian don't.

**Comment by philmmanjaro** [comment]
@aliveafter1000: can you base your code on the current released [v3.3.0](https://github.com/KDE/kpmcore/blob/3f562ab0f04ff2cc3d1e7ab3707919557650bec8/src/core/partitiontable.h#L71)? There we still have **FlagEsp**.
